cpp/include/llio.h

```diff
@@ -157,19 +156,44 @@ class message_builder {
     }
   }
 
-  void add_raw(const char *b, uint32_t size) noexcept {
-    _buffer.write(b, size);
+  void add_raw(const char *b, uint32_t len) noexcept {
+      uint free = _size - _pos;
+      if(free < len){
```

Comment:
If this resize path is not hit too frequently, a `std::vector<char>` should make things simpler: no need to track resources / take care of copying.